File: src/main/java/net/guneyilmaz0/permissionss/permission/PermissionManager.java
package net.guneyilmaz0.permissionss.permission;

import cn.nukkit.Player;
import cn.nukkit.permission.PermissionAttachment;
import net.guneyilmaz0.permissionss.Main;
import net.guneyilmaz0.permissionss.Profile;
import net.guneyilmaz0.permissionss.enums.Process;
import net.guneyilmaz0.permissionss.group.Group;
import net.guneyilmaz0.permissionss.group.GroupManager;

import java.util.HashMap;
import java.util.List;
import java.util.UUID;

public class PermissionManager {
    public static HashMap<UUID, PermissionAttachment> permissions = new HashMap<>();

    public static void addPermsToOnlinePlayers() {
        for (Player player : Main.getInstance().getServer().getOnlinePlayers().values())
            if (player != null)
                handlePermissions(player);
    }

    public static void handlePermissions(Player player) {
        PermissionAttachment attachment = player.addAttachment(Main.getInstance());
        Group group = GroupManager.getPlayerGroup(player);
        for (String perm : group.getAllPermissions()) attachment.setPermission(perm, true);

        List<String> specialPermissions = Profile.getProfile(player.getName()).getPermissions();
        for (String specialPermission : specialPermissions) attachment.setPermission(specialPermission, true);

        permissions.put(player.getUniqueId(), attachment);
    }

    public static Process addPermission(String name, String permission) {
        Profile profile = Profile.getProfile(name.toLowerCase());
        List<String> specialPermissions = profile.getPermissions();
        if (specialPermissions.contains(permission)) return Process.ALREADY_EXISTS;

        specialPermissions.add(permission);
        profile.setPermissions(specialPermissions);
        Profile.save(name, profile);
        reloadPermissions();
        return Process.SUCCESS;
    }

    public static Process removePermission(String name, String permission) {
        Profile profile = Profile.getProfile(name.toLowerCase());
        List<String> specialPermissions = profile.getPermissions();
        if (!specialPermissions.contains(permission)) return Process.NOT_FOUND;

        specialPermissions.remove(permission);
        profile.setPermissions(specialPermissions);
        Profile.save(name, profile);
        reloadPermissions();
        return Process.SUCCESS;
    }

    public static void removePermissions() {
        for (Player player : Main.getInstance().getServer().getOnlinePlayers().values()) {
            PermissionAttachment attachment = permissions.get(player.getUniqueId());
            player.removeAttachment(attachment);
        }
    }

    public static void reloadPermissions() {
        removePermissions();
        addPermsToOnlinePlayers();
    }
}

File: src/main/kotlin/permission/PermissionManager.kt
package net.guneyilmaz0.permissions.permission

import cn.nukkit.Player
import cn.nukkit.Server
import cn.nukkit.permission.PermissionAttachment
import net.guneyilmaz0.permissions.enums.Process
import net.guneyilmaz0.permissions.PermissionsS
import net.guneyilmaz0.permissions.group.Group
import net.guneyilmaz0.permissions.group.GroupManager
import java.util.*

object PermissionManager {
    val attachments: HashMap<UUID, PermissionAttachment> = HashMap()

    fun addPermsToOnlinePlayers() {
        for (value in PermissionsS.instance.server.onlinePlayers.values) handlePermissions(value)
    }

    fun handlePermissions(player: Player) {
        if (attachments.containsKey(player.uniqueId)) player.removeAttachment(attachments[player.uniqueId]!!)

        val attachment = player.addAttachment(PermissionsS.instance)
        val group: Group = GroupManager.getPlayerGroup(player)

        group.getAllPermissions().forEach { perm ->
            attachment.setPermission(perm, true)
        }

        if (!PermissionsS.provider.isProfileExists(player.uniqueId)) return

        val specialPermissions: List<String> = PermissionsS.provider.getProfile(player.uniqueId)?.permissions ?: emptyList()
        specialPermissions.forEach { specialPermission ->
            attachment.setPermission(specialPermission, true)
        }

        attachments[player.uniqueId] = attachment
    }

    @JvmStatic
    fun addPermission(uuid: UUID, permission: String): Process {
        if (!PermissionsS.provider.isProfileExists(uuid)) return Process.NOT_FOUND
        val profile = PermissionsS.provider.getProfile(uuid)!!
        val specialPermissions = profile.permissions.toMutableList()

        if (specialPermissions.contains(permission)) return Process.ALREADY_EXISTS

        specialPermissions.add(permission)
        profile.permissions = specialPermissions
        PermissionsS.provider.saveProfile(profile)
        val player = Server.getInstance().getOfflinePlayer(uuid)
        if (player.isOnline) handlePermissions(player as Player)
        return Process.SUCCESS
    }

    @JvmStatic
    fun removePermission(uuid: UUID, permission: String): Process {
        if (!PermissionsS.provider.isProfileExists(uuid)) return Process.NOT_FOUND
        val profile = PermissionsS.provider.getProfile(uuid)!!
        val specialPermissions = profile.permissions.toMutableList()

        if (!specialPermissions.contains(permission)) return Process.NOT_FOUND

        specialPermissions.remove(permission)
        profile.permissions = specialPermissions
        PermissionsS.provider.saveProfile(profile)
        val player = Server.getInstance().getOfflinePlayer(uuid)
        if (player.isOnline) handlePermissions(player as Player)
        return Process.SUCCESS
    }
}
